File: packages/catalyst-core/src/native/androidProject/app/src/main/java/io/yourname/androidproject/utils/PerfEventBuffer.kt
package io.yourname.androidproject.utils

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.webkit.WebView
import io.yourname.androidproject.BuildConfig
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicBoolean

/**
 * PerfEventBuffer — native ring buffer for Catalyst profiler events.
 *
 * Collects all events from onCreate (boot timing, cache, API calls) and flushes
 * them as a single JSON array to the web layer after onPageFinished.
 *
 * Design:
 *   - add() is safe from any thread (ConcurrentLinkedQueue)
 *   - scheduleFlush() called from onPageFinished — posts to main thread after 250ms
 *     so the JS environment is ready to receive events
 *   - flush() serialises all events to JSON array, calls window.__catalystPerfBatch()
 *   - CacheSummarizer: running tally of hit/miss/fetch + top-5 min-heap for InsightsCollector
 *   - Max ring buffer size: 500 events (oldest dropped if exceeded)
 *
 * API call timing:
 *   - bridgeCallReceived(callId, method) → stamps t0, stores in pendingCalls map
 *   - bridgeCallDispatched(callId) → stamps t1, computes durationMs, adds api-call event to buffer
 */
object PerfEventBuffer {

    private const val TAG = "PerfEventBuffer"
    private const val MAX_BUFFER_SIZE = 500
    private const val FLUSH_DELAY_MS = 250L
    private const val FLUSH_RETRY_DELAY_MS = 250L
    private const val MAX_FLUSH_RETRIES = 5
    private const val PERIODIC_FLUSH_INTERVAL_MS = 4000L

    private val buffer = ConcurrentLinkedQueue<JSONObject>()
    private val mainHandler = Handler(Looper.getMainLooper())
    private val flushed = AtomicBoolean(false)
    @Volatile private var profilerEnabled = false
    private val deliveryState = PerfDeliveryState<JSONArray>()

    // Periodic flush state
    @Volatile private var periodicWebView: WebView? = null
    private var lastFlushedSize = 0
    private val periodicRunnable: Runnable = object : Runnable {
        override fun run() {
            if (!isEnabled()) return
            val wv = periodicWebView ?: return
            val pendingBatch = deliveryState.current()
            if (pendingBatch != null) {
                deliver(wv, pendingBatch)
            } else if (buffer.size > lastFlushedSize) {
                val currentSize = buffer.size
                android.util.Log.d(TAG, "[PerfBuffer] periodicFlush() — $currentSize events since last flush")
                beginFlush(wv, includeCacheSummary = false)
            }
            mainHandler.postDelayed(this, PERIODIC_FLUSH_INTERVAL_MS)
        }
    }

    // Pending API call timing: callId → t0 nativeMs + method
    private val pendingCalls = java.util.concurrent.ConcurrentHashMap<String, Pair<Long, String>>()

    // ─── CacheSummarizer ─────────────────────────────────────────────────────

    private var cacheHits   = 0
    private var cacheMisses = 0
    private var cacheFetches = 0
    private var cacheTotalMs = 0L
    // top-5 slowest: list of Pair(durationMs, filename)
    private val cacheTopSlow = java.util.PriorityQueue<Pair<Long, String>>(
        6, compareBy { it.first }
    )

    private fun updateCacheSummary(type: String, durationMs: Long, url: String) {
        val filename = try {
            val path = java.net.URI(url).path
            path.split("/").filter { it.isNotEmpty() }.lastOrNull()?.take(40) ?: url.take(40)
        } catch (e: Exception) { url.take(40) }

        when (type) {
            "cache-hit-memory", "cache-hit-disk" -> cacheHits++
            "cache-miss-fetch"                   -> cacheMisses++
            "network-fetch-complete"             -> {
                cacheFetches++
                cacheTotalMs += durationMs
                cacheTopSlow.add(Pair(durationMs, filename))
                if (cacheTopSlow.size > 5) cacheTopSlow.poll() // remove smallest
            }
        }
    }

    private fun buildCacheSummaryEvent(): JSONObject? {
        val total = cacheHits + cacheMisses
        if (total == 0) return null
        val hitRate = if (total > 0) (cacheHits * 100.0 / total) else 0.0
        val avgFetchMs = if (cacheFetches > 0) cacheTotalMs / cacheFetches else 0L
        val topSlow = JSONArray()
        cacheTopSlow.sortedByDescending { it.first }.forEach { (ms, name) ->
            topSlow.put(JSONObject().apply {
                put("filename", name)
                put("durationMs", ms)
            })
        }
        return JSONObject().apply {
            put("type", "cache-summary")
            put("hits", cacheHits)
            put("misses", cacheMisses)
            put("fetches", cacheFetches)
            put("total", total)
            put("hitRatePct", Math.round(hitRate * 10) / 10.0)
            put("avgFetchMs", avgFetchMs)
            put("topSlowest", topSlow)
            put("nativeTime", SystemClock.elapsedRealtime())
        }
    }

    // ─── Public API ──────────────────────────────────────────────────────────

    fun configure(enabled: Boolean) {
        profilerEnabled = BuildConfig.DEBUG && enabled
        if (!profilerEnabled) reset()
    }

    fun isEnabled(): Boolean = profilerEnabled

    /** Add a perf event to the buffer. Safe from any thread. */
    fun add(event: JSONObject) {
        if (!isEnabled()) return
        if (buffer.size >= MAX_BUFFER_SIZE) {
            buffer.poll() // drop oldest
        }
        // Update cache summary tally
        val type = event.optString("type")
        if (type.startsWith("cache-") || type == "network-fetch-complete") {
            updateCacheSummary(type, event.optLong("durationMs", 0), event.optString("url"))
        }
        buffer.add(event)
        android.util.Log.d(TAG, "[PerfBuffer] add() type=$type bufferSize=${buffer.size}")
    }

    /**
     * Record start of an @JavascriptInterface bridge call.
     * Call this immediately when the method is invoked on the JavascriptInterface.
     */
    fun bridgeCallReceived(callId: String, method: String) {
        if (!isEnabled()) return
        pendingCalls[callId] = Pair(SystemClock.elapsedRealtime(), method)
    }

    /**
     * Record the end of immediate native bridge-handler work and add an api-call event.
     */
    fun bridgeCallDispatched(callId: String) {
        if (!isEnabled()) return
        val pending = pendingCalls.remove(callId) ?: return
        val (t0, method) = pending
        val t1 = SystemClock.elapsedRealtime()
        val durationMs = t1 - t0
        add(JSONObject().apply {
            put("type", "api-call")
            put("callId", callId)
            put("method", method)
            put("nativeStartMs", t0)
            put("nativeEndMs", t1)
            put("durationMs", durationMs)
            put("thread", Thread.currentThread().name)
        })
    }

    /**
     * Schedule a flush after onPageFinished (initial/full page load).
     * Also starts the periodic flush loop for subsequent SPA navigations.
     */
    fun scheduleFlush(webView: WebView) {
        if (!isEnabled()) return
        if (flushed.getAndSet(true)) {
            android.util.Log.d(TAG, "[PerfBuffer] scheduleFlush() skipped — already flushed")
            return
        }
        android.util.Log.d(TAG, "[PerfBuffer] scheduleFlush() scheduled in ${FLUSH_DELAY_MS}ms, bufferSize=${buffer.size}")
        mainHandler.postDelayed({
            beginFlush(webView, includeCacheSummary = true)
            // Start periodic flush loop for SPA navigations after initial load
            startPeriodicFlush(webView)
        }, FLUSH_DELAY_MS)
    }

    /**
     * Flush any pending events immediately (e.g. before a full URL navigation).
     * Does NOT reset the buffer — only drains buffered events up to this point.
     */
    fun flushNow(webView: WebView) {
        if (!isEnabled()) return
        mainHandler.post { beginFlush(webView, includeCacheSummary = false) }
    }

    /**
     * Start periodic flush loop — fires every 4s and flushes only if new events exist.
     * Call after onPageFinished so the JS receiver is registered.
     */
    fun startPeriodicFlush(webView: WebView) {
        if (!isEnabled()) return
        mainHandler.removeCallbacks(periodicRunnable)
        periodicWebView = webView
        lastFlushedSize = 0
        mainHandler.postDelayed(periodicRunnable, PERIODIC_FLUSH_INTERVAL_MS)
        android.util.Log.d(TAG, "[PerfBuffer] startPeriodicFlush() — interval=${PERIODIC_FLUSH_INTERVAL_MS}ms")
    }

    /** Stop periodic flush (call on destroy / loadUrl reset). */
    fun stopPeriodicFlush() {
        mainHandler.removeCallbacks(periodicRunnable)
        periodicWebView = null
    }

    /**
     * Reset the buffer for the next page load.
     * Call this from loadUrl so boot timing events for the new page are captured.
     */
    fun reset() {
        stopPeriodicFlush()
        buffer.clear()
        deliveryState.reset()
        pendingCalls.clear()
        flushed.set(false)
        lastFlushedSize = 0
        cacheHits = 0
        cacheMisses = 0
        cacheFetches = 0
        cacheTotalMs = 0L
        cacheTopSlow.clear()
    }

    // ─── Flush ───────────────────────────────────────────────────────────────

    /** Moves the current queue into an immutable in-flight batch until JS acknowledges it. */
    private fun beginFlush(webView: WebView, includeCacheSummary: Boolean) {
        if (!isEnabled()) return
        if (Looper.myLooper() != Looper.getMainLooper()) {
            mainHandler.post { beginFlush(webView, includeCacheSummary) }
            return
        }
        if (deliveryState.hasInFlight()) return
        val events = JSONArray()
        while (buffer.isNotEmpty()) {
            buffer.poll()?.let { events.put(it) }
        }
        if (includeCacheSummary) buildCacheSummaryEvent()?.let { events.put(it) }
        if (events.length() == 0) return

        val batch = deliveryState.begin(events) ?: return
        deliver(webView, batch)
    }

    private fun deliver(webView: WebView, batch: PerfDeliveryState.Batch<JSONArray>) {
        if (!isEnabled() || !deliveryState.isCurrent(batch)) return
        try {
            val json = batch.payload.toString().replace("\\", "\\\\").replace("'", "\\'")
            val batchId = JSONObject.quote(batch.id)
            webView.evaluateJavascript(
                "window.__catalystPerfBatch ? window.__catalystPerfBatch($batchId, '$json') : false",
            ) { result ->
                if (result == "true") acknowledge(batch) else retry(webView, batch)
            }
        } catch (_: Exception) {
            retry(webView, batch)
        }
    }

    private fun acknowledge(batch: PerfDeliveryState.Batch<JSONArray>) {
        if (deliveryState.acknowledge(batch)) {
            lastFlushedSize = 0
        }
    }

    private fun retry(webView: WebView, batch: PerfDeliveryState.Batch<JSONArray>) {
        val retry = deliveryState.retry(batch, MAX_FLUSH_RETRIES)
        if (retry == null) {
            android.util.Log.w(TAG, "[PerfBuffer] batch ${batch.id} was not acknowledged; retaining until reset")
            return
        }
        mainHandler.postDelayed({ deliver(webView, retry) }, FLUSH_RETRY_DELAY_MS)
    }
}

File: packages/catalyst-core/src/native/iosnativeWebView/Sources/Core/Utils/PerfEventBuffer.swift
import Foundation
import WebKit
import MachO

enum CatalystPerf {
    private static let maxBufferSize = 500
    private static let flushDelayMs: UInt64 = 250
    private static let periodicFlushIntervalMs: UInt64 = 4_000
    private static let lock = NSLock()

    private static var buffer: [[String: Any]] = []
    private static var pendingCalls: [String: (startMs: Int64, method: String)] = [:]
    private static var flushed = false
    private static weak var periodicWebView: WKWebView?
    private static var periodicTask: Task<Void, Never>?

    private static var cacheHits = 0
    private static var cacheMisses = 0
    private static var cacheFetches = 0
    private static var cacheTotalMs: Int64 = 0
    private static var cacheTopSlow: [(durationMs: Int64, filename: String)] = []

    static func nativeTimeMs() -> Int64 {
        Int64(ProcessInfo.processInfo.systemUptime * 1000)
    }

    static func add(_ event: [String: Any]) {
        #if DEBUG
        guard ConfigConstants.Profiler.enabled else { return }
        var payload = event
        if payload["nativeTime"] == nil {
            payload["nativeTime"] = nativeTimeMs()
        }

        lock.lock()
        if buffer.count >= maxBufferSize {
            buffer.removeFirst()
        }
        updateCacheSummaryLocked(payload)
        buffer.append(payload)
        lock.unlock()
        #endif
    }

    static func emit(_ event: [String: Any], to webView: WKWebView?) {
        #if DEBUG
        guard ConfigConstants.Profiler.enabled else { return }
        guard let webView else {
            add(event)
            return
        }

        var payload = event
        if payload["nativeTime"] == nil {
            payload["nativeTime"] = nativeTimeMs()
        }

        DispatchQueue.main.async {
            guard let json = jsonString(payload) else { return }
            let script = "window.__catalystPerfEvent && window.__catalystPerfEvent(\(javaScriptStringLiteral(json)));"
            webView.evaluateJavaScript(script)
        }
        #endif
    }

    static func bridgeCallReceived(callId: String, method: String) {
        #if DEBUG
        guard ConfigConstants.Profiler.enabled else { return }
        lock.lock()
        pendingCalls[callId] = (nativeTimeMs(), method)
        lock.unlock()
        #endif
    }

    static func bridgeCallDispatched(callId: String) {
        #if DEBUG
        guard ConfigConstants.Profiler.enabled else { return }
        lock.lock()
        guard let pending = pendingCalls.removeValue(forKey: callId) else {
            lock.unlock()
            return
        }
        lock.unlock()

        let endMs = nativeTimeMs()
        add([
            "type": "api-call",
            "callId": callId,
            "method": pending.method,
            "nativeStartMs": pending.startMs,
            "nativeEndMs": endMs,
            "durationMs": endMs - pending.startMs,
            "thread": Thread.current.name ?? "unknown",
        ])
        #endif
    }

    static func scheduleFlush(_ webView: WKWebView?) {
        #if DEBUG
        guard ConfigConstants.Profiler.enabled else { return }
        guard let webView else { return }

        lock.lock()
        if flushed {
            lock.unlock()
            return
        }
        flushed = true
        lock.unlock()

        Task { @MainActor in
            try? await Task.sleep(nanoseconds: flushDelayMs * 1_000_000)
            flush(webView, includeCacheSummary: true)
            startPeriodicFlush(webView)
        }
        #endif
    }

    static func flushNow(_ webView: WKWebView?) {
        #if DEBUG
        guard ConfigConstants.Profiler.enabled else { return }
        guard let webView else { return }
        Task { @MainActor in
            flush(webView, includeCacheSummary: false)
        }
        #endif
    }

    static func reset() {
        #if DEBUG
        stopPeriodicFlush()
        lock.lock()
        buffer.removeAll()
        pendingCalls.removeAll()
        flushed = false
        cacheHits = 0
        cacheMisses = 0
        cacheFetches = 0
        cacheTotalMs = 0
        cacheTopSlow.removeAll()
        lock.unlock()
        #endif
    }

    static func memorySnapshot(to webView: WKWebView?, label: String? = nil) {
        #if DEBUG
        guard ConfigConstants.Profiler.enabled else { return }
        let currentMb = currentResidentMemoryMb()
        let event: [String: Any] = [
            "type": "memory-snapshot",
            "nativeTime": nativeTimeMs(),
            "jvmMb": 0,
            "webviewMb": currentMb,
            "nativeMb": currentMb,
            "totalMb": currentMb,
            "otherMb": 0,
            "peakMb": currentMb,
            "thread": Thread.current.name ?? "unknown",
            "label": label ?? "ios-process",
        ]

        if let webView {
            emit(event, to: webView)
        } else {
            add(event)
        }
        #endif
    }

    static func injectNativeTimeOffset(into webView: WKWebView?) {
        #if DEBUG
        guard ConfigConstants.Profiler.enabled else { return }
        guard let webView else { return }
        let nativeNow = nativeTimeMs()
        DispatchQueue.main.async {
            webView.evaluateJavaScript("window.__NATIVE_TIME_OFFSET = \(nativeNow) - performance.now();")
        }
        #endif
    }

    private static func startPeriodicFlush(_ webView: WKWebView) {
        periodicTask?.cancel()
        periodicWebView = webView
        periodicTask = Task { @MainActor in
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: periodicFlushIntervalMs * 1_000_000)
                guard let webView = periodicWebView else { return }
                flush(webView, includeCacheSummary: false)
                memorySnapshot(to: webView, label: "periodic")
            }
        }
    }

    private static func stopPeriodicFlush() {
        periodicTask?.cancel()
        periodicTask = nil
        periodicWebView = nil
    }

    @MainActor
    private static func flush(_ webView: WKWebView, includeCacheSummary: Bool) {
        lock.lock()
        var events = buffer
        buffer.removeAll()
        if includeCacheSummary, let summary = buildCacheSummaryEventLocked() {
            events.append(summary)
        }
        lock.unlock()

        guard !events.isEmpty, let json = jsonString(events) else { return }
        let script = "window.__catalystPerfBatch && window.__catalystPerfBatch(\(javaScriptStringLiteral(json)));"
        webView.evaluateJavaScript(script)
    }

    private static func updateCacheSummaryLocked(_ event: [String: Any]) {
        guard let type = event["type"] as? String else { return }
        let durationMs = event["durationMs"] as? Int64 ?? Int64(event["durationMs"] as? Int ?? 0)
        let url = event["url"] as? String ?? ""

        switch type {
        case "cache-hit-memory", "cache-hit-disk":
            cacheHits += 1
        case "cache-miss-fetch":
            cacheMisses += 1
        case "network-fetch-complete":
            cacheFetches += 1
            cacheTotalMs += durationMs
            cacheTopSlow.append((durationMs, filename(from: url)))
            cacheTopSlow = Array(cacheTopSlow.sorted { $0.durationMs > $1.durationMs }.prefix(5))
        default:
            break
        }
    }

    private static func buildCacheSummaryEventLocked() -> [String: Any]? {
        let total = cacheHits + cacheMisses
        guard total > 0 else { return nil }
        let hitRate = Double(cacheHits) * 100.0 / Double(total)
        let avgFetchMs = cacheFetches > 0 ? cacheTotalMs / Int64(cacheFetches) : 0
        return [
            "type": "cache-summary",
            "hits": cacheHits,
            "misses": cacheMisses,
            "fetches": cacheFetches,
            "total": total,
            "hitRatePct": (hitRate * 10).rounded() / 10,
            "avgFetchMs": avgFetchMs,
            "topSlowest": cacheTopSlow.map { ["filename": $0.filename, "durationMs": $0.durationMs] },
            "nativeTime": nativeTimeMs(),
        ]
    }

    private static func currentResidentMemoryMb() -> Int {
        var info = mach_task_basic_info()
        var count = mach_msg_type_number_t(MemoryLayout<mach_task_basic_info>.size) / 4
        let result = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                task_info(mach_task_self_, task_flavor_t(MACH_TASK_BASIC_INFO), $0, &count)
            }
        }
        guard result == KERN_SUCCESS else { return 0 }
        return Int(info.resident_size / 1024 / 1024)
    }

    private static func filename(from url: String) -> String {
        URL(string: url)?.lastPathComponent.prefix(40).description ?? String(url.prefix(40))
    }

    private static func jsonString(_ object: Any) -> String? {
        guard JSONSerialization.isValidJSONObject(object),
              let data = try? JSONSerialization.data(withJSONObject: object),
              let json = String(data: data, encoding: .utf8) else {
            return nil
        }
        return json
    }

    private static func javaScriptStringLiteral(_ value: String) -> String {
        let escaped = value
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "'", with: "\\'")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "\\r")
            .replacingOccurrences(of: "\u{2028}", with: "\\u2028")
            .replacingOccurrences(of: "\u{2029}", with: "\\u2029")
        return "'\(escaped)'"
    }
}
